---
title: "Class 13: RNASeq Mini Project"
author: "Benjamin lee"
format: pdf
---

The data for for hands-on session comes from GEO entry: GSE37704, which is associated with the following publication:

Trapnell C, Hendrickson DG, Sauvageau M, Goff L et al. "Differential analysis of gene regulation at transcript resolution with RNA-seq". Nat Biotechnol 2013 Jan;31(1):46-53. PMID: 23222703

## RNASeq input data 

Again I need two things 

-countData
-colData

```{r}
colData <- read.csv("GSE37704_metadata.csv", row.names=1)
head(colData)
```

```{r}
countData <- read.csv("GSE37704_featurecounts.csv", row.names=1)
head(countData)
```

There is an unwanted first column "length" in the countData. I will need to remove this first before going on to further analysis.

```{r}
counts <- countData[,-1]
head(counts)
```

```{r}
all(colnames(counts) == rownames(colData))
```

## Remove zero count genes
There are lots of genes here with no count data - i.e. zero counts in all experiments. Let's remove these before running DESeq
```{r}
head(counts)
```

```{r}
to.keep <- rowSums(counts) > 0
counts <- counts[to.keep,]
head(counts)
```

How many genes do we have left?
```{r}
nrow(counts)
```


Time to use DESeq
```{r}
library(DESeq2)
```

1st step: Setup the object required by DESeq
```{r}
dds <- DESeqDataSetFromMatrix(countData = counts,
                       colData = colData,
                       design = ~condition)
```

Run the analysis
```{r}
dds <- DESeq(dds)
res <- results(dds)
```

```{r}
head(res)
```

# Volcano plot

```{r}
plot(res$log2FoldChange, log(res$padj))
```

I want to add some color. Take a fold-change threshold of -2/+2 and an alpha p-adj (p-value) threshold of 0.05
```{r}
mycols <- rep("gray", nrow(counts))
mycols[ abs(res$log2FoldChange) > 2 ] <- "blue"
mycols[ res$padj > 0.05 ] <- "red"

plot( res$log2FoldChange, -log(res$padj), col=mycols)
abline(v=c(-2,+2))
abline(h=-log(0.05))

```

# Adding gene annotation

I am going to add the database identiifiers I need for pathway analysis here

```{r}
library("AnnotationDbi")
library("org.Hs.eg.db")

columns(org.Hs.eg.db)
```

```{r}
res$symbol = mapIds(org.Hs.eg.db,
                    keys=rownames(res), 
                    keytype="ENSEMBL",
                    column="SYMBOL",
                    multiVals="first")

res$entrez = mapIds(org.Hs.eg.db,
                    keys=rownames(res),
                    keytype="ENSEMBL",
                    column="ENTREZID",
                    multiVals="first")
head(res)

```

Save my results so far to a CSV file 
```{r}
res = res[order(res$pvalue),]
write.csv(res, file="deseq_results.csv")
```


## Pathway Analysis

Again we will use the 'gage()' package and function with a focus first on KEGG and GO. 

```{r}
library(gage)
library(gageData)

data(kegg.sets.hs)
data(sigmet.idx.hs)

# Focus on signaling and metabolic pathways only
kegg.sets.hs = kegg.sets.hs[sigmet.idx.hs]

```

Recall that 'gage()' function wants only a vector of importance as the input that has names in ENTREZID format.
```{r}
foldchanges = res$log2FoldChange
names(foldchanges) = res$entrez
head(foldchanges)
```

```{r}
# Get the results
keggres = gage(foldchanges, gsets=kegg.sets.hs)
```

```{r}
head(keggres$less, 5)
```

Generate a colored pathway figure for hsa04110 Cell cycle 

```{r}
library(pathview)
pathview(gene.data=foldchanges, pathway.id="hsa04110")
```

![](hsa04110.pathview.png)

## Overall Steps

Read colData & countData 
-check data
-filter zero count genes

Run DESeq
-Plot 
focus on abs (fold-change) and padj (p-value)

Annotation

Pathway Analysis 
- KEGG, GO, etc. 

## Gene Ontology (GO)

We can also do a similar procedure with gene ontology. Similar to above, go.sets.hs has all GO terms. go.subs.hs is a named list containing indexes for the BP, CC, and MF ontologies. Let’s focus on BP (a.k.a Biological Process) here.

```{r}
data(go.sets.hs)
data(go.subs.hs)

# Focus on Biological Process subset of GO
gobpsets = go.sets.hs[go.subs.hs$BP]

gobpres = gage(foldchanges, gsets=gobpsets, same.dir=TRUE)

lapply(gobpres, head)
```


## Reactome Analysis 

Reactome is database consisting of biological molecules and their relation to pathways and processes.
```{r}
sig_genes <- res[res$padj <= 0.05 & !is.na(res$padj), "symbol"]
print(paste("Total number of significant genes:", length(sig_genes)))
```

```{r}
write.table(sig_genes, file="significant_genes.txt", row.names=FALSE, col.names=FALSE, quote=FALSE)
```

>Q: What pathway has the most significant “Entities p-value”? Do the most significant pathways listed match your previous KEGG results? What factors could cause differences between the two methods?


```{r}
# most significant pathway with regards to entities p-value
min(sig_genes, na.rm = TRUE)
# we can see that most significant pathways dont match KEGG results
all(sig_genes == gobpres)
```

## GO Optional 
```{r}
sessionInfo()
```

